File: module_widget/src/main/java/com/yyxnb/module_widget/viewmodel/MainViewModel.java
package com.yyxnb.module_widget.viewmodel;

import android.arch.paging.DataSource;
import android.arch.paging.PageKeyedDataSource;
import android.support.annotation.NonNull;

import com.yyxnb.module_widget.bean.MainBean;
import com.yyxnb.module_widget.config.DataConfig;
import com.yyxnb.common_base.core.BasePagedViewModel;

import java.util.Collections;

public class MainViewModel extends BasePagedViewModel<MainBean> {
    @Override
    public DataSource createDataSource() {
        return new PageKeyedDataSource<Integer, MainBean>() {
            @Override
            public void loadInitial(@NonNull LoadInitialParams<Integer> params, @NonNull LoadInitialCallback<Integer, MainBean> callback) {
                callback.onResult(DataConfig.getMainBeans(), null, null);
            }

            @Override
            public void loadBefore(@NonNull LoadParams<Integer> params, @NonNull LoadCallback<Integer, MainBean> callback) {
                callback.onResult(Collections.emptyList(), null);
            }

            @Override
            public void loadAfter(@NonNull LoadParams<Integer> params, @NonNull LoadCallback<Integer, MainBean> callback) {
                callback.onResult(Collections.emptyList(), null);
            }
        };
    }

    //invalidate 之后Paging会重新创建一个DataSource 重新调用它的loadInitial方法加载初始化数据
    public void invalidateDataSource() {
        getDataSource().invalidate();
    }

}

File: module_main/src/main/java/com/yyxnb/module_main/viewmodel/MainViewModel.kt
package com.yyxnb.module_main.viewmodel

import androidx.paging.DataSource
import androidx.paging.PageKeyedDataSource
import com.yyxnb.module_main.bean.MainHomeBean
import com.yyxnb.module_main.config.DataConfig.mainBeans
import com.yyxnb.network.BasePagedViewModel

class MainViewModel : BasePagedViewModel<MainHomeBean>() {

    override fun createDataSource(): DataSource<*, *> {
        return object : PageKeyedDataSource<Int?, MainHomeBean?>() {
            override fun loadInitial(params: LoadInitialParams<Int?>, callback: LoadInitialCallback<Int?, MainHomeBean?>) {
                callback.onResult(mainBeans!!, null, null)
            }

            override fun loadBefore(params: LoadParams<Int?>, callback: LoadCallback<Int?, MainHomeBean?>) {
                callback.onResult(emptyList(), null)
            }

            override fun loadAfter(params: LoadParams<Int?>, callback: LoadCallback<Int?, MainHomeBean?>) {
                callback.onResult(emptyList(), null)
            }
        }
    }
}